WebDriverAgentLib/Categories/XCUIElement+FBUtilities.m

```diff
@@ -80,6 +90,54 @@ - (XCElementSnapshot *)fb_lastSnapshot
   return self.lastSnapshot;
 }
 
+- (XCElementSnapshot *)fb_snapshotWithAttributes {
+  if (![XCElementSnapshot.class respondsToSelector:@selector(snapshotAttributesForElementSnapshotKeyPaths:)]) {
+    return nil;
+  }
+  [self resolve];
+  
+  id defaultParameters = [[XCAXClient_iOS sharedClient] defaultParameters];
+  
+  // Names of the properties to load. There won't be lazy loading for missing properties,
+  // thus missing properties will lead to wrong results
+  NSArray<NSString *> *propertyNames = @[
+                                         @"identifier",
+                                         @"value",
+                                         @"label",
+                                         @"frame",
+                                         @"enabled",
+                                         @"elementType"
+                                         ];
+  
+  NSSet *attributes = [XCElementSnapshot snapshotAttributesForElementSnapshotKeyPaths:propertyNames];
+  
+  NSArray *axAttributes = XCAXAccessibilityAttributesForStringAttributes(attributes);
+  if (![axAttributes containsObject:FB_XCAXAIsVisibleAttribute]) {
```

Comment:
I wouldn't include FB_XCAXAIsVisibleAttribute attribute by default. the reason for it is that XCTest has 15-seconds hardcoded delay if we try to detect visibility on some particular element or in some particular UI states (for example running animations).

Reply:
Does this happen for any long-running animation on an UIView? And do you know if this is also the case if we detect hit-points? (Our assumption here is that, if XCTest doesn't return a hit-point, the view is not visible)

Reply:
however if this is really async and _XCT_snapshotForElement times out properly then it should not be a problem

Reply:
Yes, but in that case we're back in the state of the initial problem that the page source command can timeout. I'll check if we can go with hit-points here and see how this behaves with animations

Reply:
we only need to take a single snapshot of the root element and its subwindows to construct the page source. But yeah, hit points are needed in many cases

Reply:
> (Our assumption here is that, if XCTest doesn't return a hit-point, the view is not visible)

Not necessarily. I saw situations when hit point was set to nil for elements inside table cells. 
Also, it may return nil for elements, which are visible, but not clickable, for example covered with another transparent elements. Sometimes it even throws an exception instead of returning nil. this is very unstable and not very reliable.

You can check the details on the current snapshot visibility implementation. I was trying to cover all corner cases that I faced there.

Reply:
I created an view with an endless animation, but both ways of getting the visibility, with the snapshot attributes and via `[self fb_attributeValue:FB_XCAXAIsVisibleAttribute]`, work without any issues. Is there an edge case you've encountered where this happens?

Reply:
There is no strict pattern or at least we were not able to detect it. Check https://github.com/facebook/WebDriverAgent/issues/372